document-store/src/main/java/org/hypertrace/core/documentstore/postgres/query/v1/transformer/PostgresUnnestQueryTransformer.java

```diff
@@ -0,0 +1,267 @@
+    }
+    return finalFromTypeExpressions;
+  }
+
+  private static void updateModifiedUnnestExpression(
+      List<FromTypeExpression> finalFromTypeExpressions,
+      FromTypeExpression org,
+      FromTypeExpression modified) {
+    int foundAt =
+        IntStream.range(0, finalFromTypeExpressions.size())
+            .filter(i -> finalFromTypeExpressions.get(i).equals(org))
+            .findFirst()
+            .getAsInt();
+    finalFromTypeExpressions.set(foundAt, modified);
+  }
+
+  private static Optional<UnnestExpression> getMatchedUnnestExpression(
+      List<FromTypeExpression> fromTypeExpressions,
+      FilterTypeExpression filterTypeExpression,
+      UnnestExpressionProvider unnestExpressionProvider,
+      FilterIdentifierProvider filterIdentifierProvider) {
+    return fromTypeExpressions.stream()
+        .map(
+            fromTypeExpression ->
+                (UnnestExpression) fromTypeExpression.accept(unnestExpressionProvider))
+        .filter(
+            unnestExpression ->
+                filterTypeExpression.accept(
+                    new FilterToUnnestExpressionMatcher(
+                        filterIdentifierProvider, unnestExpression)))
+        .max(Comparator.comparingInt(u -> u.getIdentifierExpression().getName().length()));
+  }
+
+  private static UnnestExpression getBuildUnnestExpression(
+      UnnestExpression matchedUnnestExpression, FilterTypeExpression filterTypeExpression) {
+    return UnnestExpression.builder()
+        .identifierExpression(matchedUnnestExpression.getIdentifierExpression())
+        .preserveNullAndEmptyArrays(matchedUnnestExpression.isPreserveNullAndEmptyArrays())
+        .filterTypeExpression(
+            buildAndFilter(filterTypeExpression, matchedUnnestExpression.getFilterTypeExpression()))
+        .build();
+  }
+
+  @SuppressWarnings("unchecked")
+  private static class UnnestExpressionProvider implements FromTypeExpressionVisitor {
+
+    @Override
+    public UnnestExpression visit(UnnestExpression unnestExpression) {
+      return unnestExpression;
+    }
+  }
+
+  @SuppressWarnings("unchecked")
+  private static class FilterIdentifierProvider implements SelectTypeExpressionVisitor {
+
+    @Override
+    public List<String> visit(AggregateExpression expression) {
+      return expression.getExpression().accept(this);
+    }
+
+    @Override
+    public List<String> visit(ConstantExpression expression) {
+      return List.of();
+    }
+
+    @Override
+    public List<String> visit(FunctionExpression expression) {
+      return expression.getOperands().stream()
+          .map(exp -> exp.<List<String>>accept(this))
+          .flatMap(Collection::stream)
+          .collect(Collectors.toList());
+    }
+
+    @Override
+    public List<String> visit(IdentifierExpression expression) {
+      return List.of(expression.getName());
+    }
+  }
+
+  @SuppressWarnings("unchecked")
+  private static class AndFiltersCollector implements FilterTypeExpressionVisitor {
```

Comment:
Maybe we need a better naming for this class.
In the first look, it feels like it is collecting only `AndFilters` and not `OrFilters`. Perhaps `AndSplittingFiltersCollector`?

Reply:
Done with `AndSplitFiltersCollector`